core/lib/types/src/kzg.rs

```diff
@@ -0,0 +1,402 @@
+        opening_proof: Vec<u8>,
+        #[serde_as(as = "serde_with::hex::Hex")]
+        blob_proof: Vec<u8>,
+        #[serde_as(as = "serde_with::hex::Hex")]
+        pubdata_commitment: Vec<u8>,
+    }
+
+    #[serde_as]
+    #[derive(Debug, Serialize, Deserialize)]
+    struct KzgTest {
+        #[serde_as(as = "serde_with::hex::Hex")]
+        pubdata: Vec<u8>,
+        expected_outputs: ExpectedOutputs,
+    }
+
+    fn u8_repr_to_fr(bytes: &[u8]) -> Fr {
+        assert_eq!(bytes.len(), 32);
+        let mut ret = [0u64; 4];
+
+        for (i, item) in ret.iter_mut().enumerate() {
+            let mut repr = [0u8; 8];
+            let end = 32 - (8 * i);
+            let beg = 32 - (8 * (i + 1));
+            repr.copy_from_slice(&bytes[beg..end]);
+            *item = u64::from_be_bytes(repr);
+        }
+
+        Fr::from_repr(FrRepr(ret)).unwrap()
+    }
+
+    #[test]
+    fn kzg_test() {
+        let zksync_home = std::env::var("ZKSYNC_HOME").unwrap_or_else(|_| ".".into());
+        let path = std::path::Path::new(&zksync_home).join("etc/kzg_tests/kzg_test_0.json");
+        let contents = std::fs::read_to_string(path).unwrap();
+        let kzg_test: KzgTest = serde_json::from_str(&contents).unwrap();
+
+        let kzg_info = KzgInfo::new(kzg_test.pubdata);
+
+        // Verify all the fields were correctly computed
+        assert_eq!(
+            hex::encode(kzg_info.kzg_commitment),
+            hex::encode(kzg_test.expected_outputs.kzg_commitment)
+        );
+        assert_eq!(
+            hex::encode(kzg_info.opening_point),
+            hex::encode(kzg_test.expected_outputs.opening_point)
+        );
+        assert_eq!(
+            hex::encode(kzg_info.opening_value),
+            hex::encode(kzg_test.expected_outputs.opening_value)
+        );
+        assert_eq!(
+            hex::encode(kzg_info.opening_proof),
+            hex::encode(kzg_test.expected_outputs.opening_proof)
+        );
+        assert_eq!(
+            hex::encode(kzg_info.versioned_hash),
+            hex::encode(kzg_test.expected_outputs.versioned_hash)
+        );
+        assert_eq!(
+            hex::encode(kzg_info.blob_proof),
+            hex::encode(kzg_test.expected_outputs.blob_proof)
+        );
+
+        // Verify data we need for blob commitment on L1 returns the correct data
+        assert_eq!(
+            hex::encode(kzg_info.to_pubdata_commitment()),
+            hex::encode(kzg_test.expected_outputs.pubdata_commitment)
+        );
+
+        // Verify that the blob, commitment, and proofs are all valid
+        let blob = ethereum_4844_data_into_zksync_pubdata(&kzg_info.blob);
+        let mut poly = zksync_pubdata_into_monomial_form_poly(&blob);
+        fft(&mut poly);
+        bitreverse(&mut poly);
+
+        let mut commitment = G1Compressed::empty();
+        let v = commitment.as_mut();
+        v.copy_from_slice(&kzg_info.kzg_commitment);
+        let commitment = commitment.into_affine().unwrap();
```

Comment:
maybe another candidate for function/closure extraction

Reply:
Agree, esp since `v` seems to act as a temporary variables